### PR TITLE
BEAM-372 added a test that verifies if a coder consumes bytes equal t…

### PR DESCRIPTION
**Pull request by chandnisingh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [ ] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [ ] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [ ] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

…o encoded bytes

### CODE REVIEW

**Comment by chandnisingh** [comment]
Hi @dhalperi, please take a look.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 1]
```diff
@@ -322,6 +323,26 @@
     }
   }
 
+  public static <T> void coderConsumesBytesEqualToBytesEncoded(Coder<T> coder, T value)
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: We could update all our existing users automatically if we fold this logic into the decode(Coder<T>, Context, bytes[]) method.

There we can check to see if the context is nested and then pad the byte array before decoding.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 2]
```diff
@@ -322,6 +323,26 @@
...
+
+    @SuppressWarnings("unchecked")
+    Coder<T> deserializedCoder = Serializer.deserialize(coder.asCloudObject(), Coder.class);
+
+    byte[] extendedBytes = new byte[bytes.length + Ints.BYTES];
+    System.arraycopy(bytes, 0, extendedBytes, 0, bytes.length);
+    System.arraycopy(Ints.toByteArray(1), 0, extendedBytes, bytes.length, Ints.BYTES);
+
+    ByteArrayInputStream is = new ByteArrayInputStream(extendedBytes);
+    int startPos = is.available();
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: available doesn't need to tell the truth, better to rely on a CountingInputStream to record how many bytes were actually read.

https://google.github.io/guava/releases/19.0/api/docs/com/google/common/io/CountingInputStream.html

File: sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java [hunk 3]
```diff
@@ -322,6 +323,26 @@
     }
   }
 
+  public static <T> void coderConsumesBytesEqualToBytesEncoded(Coder<T> coder, T value)
+      throws IOException {
+
+    byte[] bytes = encode(coder, Coder.Context.NESTED, value);
+
+    @SuppressWarnings("unchecked")
+    Coder<T> deserializedCoder = Serializer.deserialize(coder.asCloudObject(), Coder.class);
```
Comment (lukecwik) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/testing/CoderProperties.java]: We don't need to serialize/deserialize the coder if this gets folded into the decode(Coder, Context, bytes[]) method

File: sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CoderPropertiesTest.java [hunk 4]
```diff
@@ -213,4 +215,39 @@ public void testClosingCoderFailsWhenEncoding() throws Exception {
...
     CoderProperties.encode(new ClosingCoder(), Context.NESTED, "test-value");
   }
+
+  /** Coder that consumes more bytes while decoding than required. */
+  public static class BadCoderThatConsumesMoreBytes extends NonDeterministicCoder {
+
+    @Override
+    public String decode(InputStream inStream, Context context) throws IOException {
+      String value = super.decode(inStream, context);
+      byte[] extraBytes = new byte[Ints.BYTES];
```
Comment (lukecwik) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/testing/CoderPropertiesTest.java]: No point in having this array, can call inStream.read() which is guaranteed to return a byte or -1 for EOF

**Comment by lukecwik** [comment]
R: @lukecwik

**Comment by chandnisingh** [comment]
@lukecwik I incorporated all the review comments and squashed my commits. Please have a look.

**Comment by lukecwik** [comment]
Thanks for the contribution, merging.

**Comment by lukecwik** [comment]
LGTM

**Pull request closed without merging**